sklearn/naive_bayes.py

```diff
@@ -958,3 +985,252 @@ def _joint_log_likelihood(self, X):
+    CategoricalNB()
+    >>> print(clf.predict(X[2:3]))
+    [3]
+    """
+
+    def __init__(self, alpha=1.0, fit_prior=True, class_prior=None,
+                 handle_unknown='warn'):
+        self.alpha = alpha
+        self.fit_prior = fit_prior
+        self.class_prior = class_prior
+        self.handle_unknown = handle_unknown
+
+    def fit(self, X, y, sample_weight=None):
+        """Fit Naive Bayes classifier according to X, y
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            Training vectors, where n_samples is the number of samples and
+            n_features is the number of features. Here, each feature of X is
+            assumed to be from a different categorical distribution.
+
+        y : array-like, shape = [n_samples]
+            Target values.
+
+        sample_weight : array-like, shape = [n_samples], (default=None)
+            Weights applied to individual samples (1. for unweighted).
+
+        Returns
+        -------
+        self : object
+        """
+        self._check_settings()
+        return super(CategoricalNB, self).fit(X, y,
+                                              sample_weight=sample_weight)
+
+    def partial_fit(self, X, y, classes=None, sample_weight=None):
+        """Incremental fit on a batch of samples.
+
+        This method is expected to be called several times consecutively
+        on different chunks of a dataset so as to implement out-of-core
+        or online learning.
+
+        This is especially useful when the whole dataset is too big to fit in
+        memory at once.
+
+        This method has some performance overhead hence it is better to call
+        partial_fit on chunks of data that are as large as possible
+        (as long as fitting in the memory budget) to hide the overhead.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            Training vectors, where n_samples is the number of samples and
+            n_features is the number of features. Here, each feature of X is
+            assumed to be from a different categorical distribution.
+
+        y : array-like, shape = [n_samples]
+            Target values.
+
+        classes : array-like, shape = [n_classes] (default=None)
+            List of all the classes that can possibly appear in the y vector.
+
+            Must be provided at the first call to partial_fit, can be omitted
+            in subsequent calls.
+
+        sample_weight : array-like, shape = [n_samples], (default=None)
+            Weights applied to individual samples (1. for unweighted).
+
+        Returns
+        -------
+        self : object
+        """
+        self._check_settings()
+        return (super(CategoricalNB, self)
+                .partial_fit(X, y, classes, sample_weight=sample_weight))
+
+    def _check_X(self, X):
+        return check_array(X, accept_sparse=False)
```

Comment:
Isn't it the default that `X` is a float matrix in Scikit-Learn? One would use integral floats like 1.0, 2.0, 3.0, etc. for categories but it should still be allowed that they are floats, right?

Reply:
This is our first estimator which handles categorical variables, so we need to figure out an appropriate way to represent categorical variables. It's not reasonable to represent categorical variables with floats IMO, so I think we should reject float values (even float values like 1.0, though a bit inconsistent with ``type_of_target``).
But you can leave it as it is, this won't block my approval at least now.

Reply:
@qinhanmin2014 thanks for your reply. I agree that integer would be better but I see two problems with it:
1) So far `X` in Scikit-Learn is a float matrix everywhere, so the need for the data type integer, at same point only, might confuse users.
2) When we have a `GeneralNB` in a later PR one would need a matrix `X` that encompasses float values for e.g. the `GaussianNB` and integer values for the `CategoricalNB`. Having different data types in a single matrix might be confusing, I am not sure it's even possible with vanilla NumPy. 
3) There is no way to have a missing value like `None` or `np.nan` in an integer matrix. Especially when people are using Pandas this might be problematic. Pandas has the habit to implicitly change the datatype from integer to float whenever it needs to express a missing value. Another option would be masked arrays but sadly many users are not even aware of them and relying on them would be problematic if Scikit-Learn wants to be compatible with Pandas.

My mid-term suggestion would be to provide a check within Scikit-Learn that assures integral float values like 1.0, 2.0, 3.0 etc. This would at least avoid mistakes where people just pass nonsense float values as categories. Long-term I would hope to see an adaption of Arrow that allows for missing integer values.

Reply:
> My mid-term suggestion would be to provide a check within Scikit-Learn that assures integral float values like 1.0, 2.0, 3.0 etc. 

It's not difficult so I guess we don't need a separate function (see sklearn.utils.validation.type_of_target: ``if y.dtype.kind == 'f' and np.any(y != y.astype(int))``).

> Long-term I would hope to see an adaption of Arrow that allows for missing integer values.

Let Naive bayse support missing value? any reference? If you have some, please open another issue.

Reply:
@qinhanmin2014, regarding the enforcement of float let's wait for the answers of the others. Maybe in some cases, when you have grades of pupils like 1.3, 2.7, 3.0, 1.0 etc. it might even be useful to allow non integral floats. 

I think a Naive Bayes estimator  (at least a categorical one) could easily support missing values by taking the probability of 1 for each feature where the category is missing. Similar to what we do right now when during inference a category pops up that wasn't seen during training. But since Scikit-Learn doesn't yet support missing values in `X` this is rather theoretical.